src/hooks/useGetTokens.ts

```diff
@@ -25,6 +34,78 @@ export function useGetTokens(networkId: Network | undefined): GetTokensResult {
     [setTokens],
   )
 
+  const setTokensHistoricalData = useCallback(
+    async (network: Network, pageIndex: number, tokenIds: string[]): Promise<void> => {
+      setIsLoading(true)
+      try {
+        const lastDayTimestamp = Number(lastHoursTimestamp(25)) // last 25 hours
+        const lastWeekTimestamp = Number(lastDaysTimestamp(8)) // last 8 days
+        const responses = {} as { [tokenId: string]: Promise<SubgraphHistoricalDataResponse> | undefined }
+        for (const tokenId of tokenIds) {
+          const response = COW_SDK[network]?.cowSubgraphApi.runQuery<SubgraphHistoricalDataResponse>(
+            GET_HISTORICAL_DATA_QUERY,
+            { address: tokenId, lastDayTimestamp, lastWeekTimestamp },
+          )
+          responses[tokenId] = response
+        }
+
+        const tokenData = {} as { [tokenId: string]: TokenData }
+
+        for (const address of Object.keys(responses)) {
+          const response = await responses[address]
+          if (!response) {
+            continue
+          }
+
+          const lastDayUsdVolume = response.tokenHourlyTotals.reduce(
+            (acc, curr) => acc + Number(curr.totalVolumeUsd),
+            0,
+          )
+          const priceUsd = response.tokenHourlyTotals[0]?.averagePrice
+
+          const lastDayTimestampFrom = Number(lastHoursTimestamp(25))
+          const lastDayTimestampTo = Number(lastHoursTimestamp(23))
+          const lastWeekTimestampFrom = Number(lastDaysTimestamp(8))
+          const lastWeekTimestampTo = Number(lastDaysTimestamp(6))
+          const lastDayPrice = response.tokenHourlyTotals.find(
+            (x) => x.timestamp >= lastDayTimestampFrom && x.timestamp <= lastDayTimestampTo,
+          )?.averagePrice
+          const lastWeekPrice = response.tokenDailyTotals.find(
+            (x) => x.timestamp >= lastWeekTimestampFrom && x.timestamp <= lastWeekTimestampTo,
+          )?.averagePrice
+
+          const prices = {
+            priceUsd,
+            lastDayPricePercentageDifference: lastDayPrice
+              ? getPercentageDifference(Number(priceUsd), Number(lastDayPrice))
+              : undefined,
+            lastWeekPricePercentageDifference: lastWeekPrice
+              ? getPercentageDifference(Number(priceUsd), Number(lastWeekPrice))
+              : undefined,
+            lastWeekUsdPrices: response.tokenDailyTotals
+              .map((x) => ({
+                time: Number(x.timestamp) as UTCTimestamp,
+                value: Number(x.averagePrice),
+              }))
+              .sort((a, b) => a.time - b.time),
+          }
+
+          tokenData[address] = { lastDayUsdVolume, ...prices }
+        }
+
+        setTokens((tokens) => [...addHistoricalData(tokens, tokenData)])
+        setHistoricalDataLoaded((loaded) => ({ ...loaded, [pageIndex]: true }))
```

Comment:
Could some sections of this function be extracted outside the hook body?
Feels like too much is going on in here and the logic could be abstracted away